### DISCUSSION TITLE
Best practice to delete (permanently) a Content Item using the Content Manager I receive in DI

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#16024] [Q&A] [answered]
Hi Everyone,

As I previously discussed [here](https://github.com/OrchardCMS/OrchardCore/discussions/16021) regarding best practices for updating/creating a Content Item using the service in DI for my API microservice, I now have a similar question about deletion.

I have noticed that the Content Manager doesn't permanently delete my `Publisher` Content Item. It only has a [RemoveAsync](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentManager.cs#L120-L126) method that unpublishes the Content Item but leaves it in the database. **I'm interested in permanently deleting the Content Item because I don't use versioning, and I prefer to keep my database clean of unused content.**

Here is how I have implemented the method in my handler class, but as mentioned, it doesn't delete the Content Item from the database:

```csharp
public async Task<ApiResult> DeletePublisherAsync(string publisherContentItemId)
{
    _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} started with publisherContentItemId: {publisherContentItemId}.");

    PublisherApiModel result = null;
    ContentItem contentItem = await _contentManager.GetAsync(publisherContentItemId);

    if (contentItem != null)
    {
        await _contentManager.RemoveAsync(contentItem);  // Unfortunately, it returns void

        // Check if the Content Item doesn't exist anymore
        ContentItem deleteResult = await _contentManager.GetAsync(publisherContentItemId);

        if (deleteResult == null)
        {
            // Deleted successfully
            _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} Publisher with publisherContentItemId: {publisherContentItemId} was deleted succesfully.");
            return new ApiResult(HttpStatusCode.OK);
        }
        else
        {
            _logger.LogError($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} An error occurred trying to delete a publisher with publisherContentItemId: {publisherContentItemId}.");
            return new ApiResult(HttpStatusCode.InternalServerError, $"An error occurred trying to delete a publisher with publisherContentItemId: {publisherContentItemId}");
        }
    }
    else
    {
        _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} No publisher found with publisherContentItemId: {publisherContentItemId}.");
        return new ApiResult(HttpStatusCode.NotFound, string.Format("No publisher found with publisherContentItemId: {id}", publisherContentItemId));
    }
}
```

Here is the `Publisher` Content Type I created in the OC backoffice, and `Versionable` is turned off:

![image](https://github.com/OrchardCMS/OrchardCore/assets/19241483/5a9a3f26-8a41-405e-984d-cf10dc5f0333)

It seems that  [this topic](https://github.com/OrchardCMS/OrchardCore/issues/6311) **was also discussed in the past**, but I can't find a solution using the `ContentManager`. 

1. Is the only way to do it using the `YesSql` [Isession.Delete method](https://github.com/sebastienros/yessql/blob/c748f4d7ae65911dd3c45c5879d3a69b86e9346e/src/YesSql.Abstractions/ISession.cs#L35-L38)? Doesn't an Orchard Core handler exist? 🤔
2. Is there a method in Orchard Core to **permanently delete** a Content Item?

Thank you

**Comment by Piedone** [comment]
This isn't supported: https://github.com/OrchardCMS/OrchardCore/issues/6056. You have to implement it low-level like the other `DefaultContentManager` CRUD operations, but at that point, it'd be better to contribute it :).

Having soft-deleted content items/versions in the DB won't become a noticeable issue until you have millions, BTW (or if you need them hard deleted for compliance reasons).

**Reply by MarGraz** [reply]
@Piedone thank you for your reply.

Yes, I need it for compliance reasons, I can't leave any "zombie" Content Items in the database.

I tried to implement the [ISession.Delete method](https://github.com/sebastienros/yessql/blob/c748f4d7ae65911dd3c45c5879d3a69b86e9346e/src/YesSql.Abstractions/ISession.cs#L35-L38), and I succeeded in deleting the element from the `Document` table, **but not from the `ContentItemIndex` table.** (*) 

This is a significant limitation for my requirements, and I need to delete data permanently.

1. Is there a way to override this behavior? Can I do it by implementing a `Query` with `ISession` in my handler?
2. **I can contribute** to this feature, **but I can't wait until the next release**. I need to implement a quick workaround first. After that, are you suggesting that I implement it in the `ContentManager`?

Thank you

> (*) _UPDATE_ it works, the problem was that I stopped the debugging too early and the process is async.

**Reply by Piedone** [reply, marked as answer]
1. I don't know; deleting the content item document should also delete the corresponding index entries.
2. Yes.

**Reply by MarGraz** [reply]
1. @Piedone I checked again, and `ISession` does delete entries in all the tables. I was confused because it's an asynchronous process, and I stopped the debugging too early 🤣
2. That could be a nice implementation, instead of directly using `ISession`.

**Because the deletion is an asynchronous process**, and `_contentManager.GetAsync` sometimes still returns the deleted item, here is the updated code, with a try/catch (because `Delete` is void, it should throw an `Exception` in case of a failure):

```csharp
public async Task<ApiResult> DeletePublisherAsync(string publisherContentItemId)
{
    _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} started with publisherContentItemId: {publisherContentItemId}.");

    PublisherApiModel result = null;
    ContentItem contentItem = await _contentManager.GetAsync(publisherContentItemId);

    if (contentItem != null)
    {
        try
        {
            _session.Delete(contentItem);
        }
        catch (Exception exc)
        {
            _logger.LogError($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} An error occurred trying to delete a publisher with publisherContentItemId: {publisherContentItemId}.");
            return new ApiResult(HttpStatusCode.InternalServerError, $"An error occurred trying to delete a publisher with publisherContentItemId: {publisherContentItemId}. Error message: {exc.Message}");
        }

        // Deleted succesfully
        _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} Publisher with publisherContentItemId: {publisherContentItemId} was deleted succesfully.");
        return new ApiResult(HttpStatusCode.OK);
    }
    else
    {
        _logger.LogInformation($"{nameof(PublisherHandler)}-{nameof(DeletePublisherAsync)} No publisher found with publisherContentItemId: {publisherContentItemId}.");
        return new ApiResult(HttpStatusCode.NotFound, string.Format("No publisher found with  publisherContentItemId: {id}", publisherContentItemId));
    }
}
```

Thank you

**Reply by Piedone** [reply]
Great then, and thanks for the update! Would you like to venture into submitting a PR? We can discuss the implementation approach before that under the issue.

**Comment by Piedone** [comment]
FYI hard delete is now available: https://github.com/OrchardCMS/OrchardCore/issues/12669
